Index: uboot-1.0-2012.07/boards.cfg
===================================================================
--- uboot-1.0-2012.07.orig/boards.cfg
+++ uboot-1.0-2012.07/boards.cfg
@@ -209,6 +209,7 @@ ipq806x_cdp                  arm
 ipq40xx_cdp                  arm         armv7       ipq40xx_cdp         qcom           qca
 wsq50                         arm         armv7       ipq40xx_cdp         qcom           qca
 wsq60                         arm         armv7       ipq40xx_cdp         qcom           qca
+wsq50c                        arm         armv7       ipq40xx_cdp         qcom           qca
 ca9x4_ct_vxp                 arm         armv7       vexpress            armltd
 am335x_evm                   arm         armv7       am335x              ti             am33xx
 highbank                     arm         armv7       highbank            -              highbank
